docs/source/blueprints/BP004.rst

```diff
@@ -0,0 +1,210 @@
+
+In a workflow considering a dependency graph, ideally, data structures related to all types of runnable should be available before creating the dependency graph. This way, the duplication of tasks for the same activity (like a requirement fulfillment) is avoided. One key factor in avoiding unneeded duplication of tasks is making the structure representing the runnable comparable.
+
+After all the data structures related to the runnables are available, the dependency graph can be built and also the topological order.
+
+The state machine will handle the same list of tasks, but this time in topological order.
+
+Internal representation
+=======================
+
+There are two ways of representing the dependency graph: implicit, using the structures already available, or explicit, creating a new graph structure.
+
+In an implicit representation, the graph arcs are defined in the runnable overlay as a doubly linked list. It is necessary to evaluate whether the doubly linked list should be part of the runnable structure, the task structure, or the runtime task structure.
+
+The advantage of this approach is that the dependency information is within the object (the runnable, task, or runtime task object). The disadvantage of this approach is that all the objects should be consulted to discover the graph structure and change it.
+
+In an explicit representation, the graph is represented in a new data structure. It can be an adjacency matrix or an adjacency list. Considering the sparsity of graphs dealt with here, the adjacency list structure should be a better choice.
+
+The advantage of an explicit representation is that the information about the graph structure is straightforward. One more data structure to handle may be considered a disadvantage, but the benefits outweigh the disadvantages.
+
+Backward Compatibility
+***********************
+
+This blueprint proposes a new feature. When it is not used, it will not affect the current behavior of the execution of the tests.
+
+As this blueprint extends the Requirements Resolver, the current behavior of a requirement is not changed.
+
+Security Implications
+*********************
+
+As far as analyzed, this feature does not introduce new security implications.
+
+It is important to note that the security implication described on BP002 for Python executable to build the requirements file is also valid in this blueprint.
+
+How to Teach This
+*****************
+
+One of the steps in developing this feature, and as soon as it replaces the current Requirements Resolver, should be to extend the `documentation <https://avocado      -framework.readthedocs.io/en/latest/guides/user/chapters/requirements.html>`__.
+
+The addition of more complex examples is another way to spread the new feature.
+
+Related Issues
+**************
+
+Here is a list of all issues related to this blueprint:
+
+#.
```

Comment:
If we have an issue about the tests that are being skipped and that would need to be enabled once this is implemented, let's add it here.

For comparison sake, BP002 had a number of issues listed here.  I'm not sure how much of the work can be foreseen and have issues created here, but if you and @richtja have an idea of the next steps it would be a good idea to add them here as issues too.

Reply:
> If we have an issue about the tests that are being skipped and that would need to be enabled once this is implemented, let's add it here.

Okay, I will do that.

> 
> For comparison sake, BP002 had a number of issues listed here. I'm not sure how much of the work can be foreseen and have issues created here, but if you and @richtja have an idea of the next steps it would be a good idea to add them here as issues too.

On BP002, we had a clearer view of what could become an issue. I opened and added all of them right after the blueprint was approved. I'll discuss with @richtja which issues we can open and add them to the blueprint as soon as the PR has all the approvals needed.

Reply:
I force-pushed with the issue that is already opened about the tasks clash.

Reply:
> 
>On BP002, we had a clearer view of what could become an issue. I opened and added all of them right after the >blueprint was approved. I'll discuss with @richtja which issues we can open and add them to the blueprint as soon as >the PR has all the approvals needed.

Yes, when this will be approved, we can discuss and create epic issue to track all the tasks which have to be done.